### PR TITLE
Implement polymorphic_allocator<>

### LINKED ISSUE #10: P0339R6 polymorphic_allocator<>
[P0339R6](https://wg21.link/P0339R6) `polymorphic_allocator<>`

LWG-3304 Allocate functions of `std::polymorphic_allocator` should require `[[nodiscard]]`

Feature-test macro as of WG21-N4842, added by WG21-P1902:
`#define __cpp_lib_polymorphic_allocator 201902L`

This paper was patched by:
LWG-3237 "LWG-3038 and LWG-3190 have inconsistent PRs (change `length_error` to `bad_array_new_length`)"

### PR DESCRIPTION
**Pull request by onihusube**
Resolves #10

References

- [P0339R6 `polymorphic_allocator<>` as a vocabulary type](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0339r6.pdf)
- [Class template polymorphic_­allocator [mem.poly.allocator.class] - N4861](https://timsong-cpp.github.io/cppwp/n4861/mem.res#mem.poly.allocator.class)

### CODE REVIEW

File: tests/std/tests/P0339R6_polymorphic_allocator/test.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    for (int i = 0; i < N; ++i) {
+        alloc.construct(arr + i, i);
+    }
+
+    for (int i = 0; i < N; ++i) {
+        assert(arr[i] == i);
+    }
+
+    for (int i = 0; i < N; ++i) {
+        std::destroy_at(arr + i);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0339R6_polymorphic_allocator/test.cpp:27]: Need to include `<memory>` for `std::destroy_at()`; `<memory_resource>` isn't guaranteed to drag this in. (That is, we try to avoid assuming the implementation's header structure in the tests, which makes it somewhat easier to restructure headers in the future, and our tests may be portable to libc++ in the distant future.)

File: tests/std/tests/P0339R6_polymorphic_allocator/test.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    alloc.deallocate_bytes(vp, N, alignof(int));
```
Comment (StephanTLavavej) [inline tests/std/tests/P0339R6_polymorphic_allocator/test.cpp:30]: :beetle: This function allocated `sizeof(int) * N` bytes, but is deallocating only `N` bytes.

File: tests/std/tests/P0339R6_polymorphic_allocator/test.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    for (int i = 0; i < N; ++i) {
+        assert(arr[i] == i);
+    }
+
+    for (int i = 0; i < N; ++i) {
+        std::destroy_at(arr + i);
+    }
+
+    alloc.deallocate_object(arr, N);
+}
```
Comment (StephanTLavavej) [inline tests/std/tests/P0339R6_polymorphic_allocator/test.cpp:53]: Should there also be test coverage for the Standard-mandated `bad_array_new_length` exception for integer overflow?

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Thanks for implementing this feature! :smile_cat: This looks solid; I have a handful of comments but they're all localized and should be straightforward to resolve.

File: tests/std/tests/P0339R6_polymorphic_allocator/test.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+#include <memory_resource>
+
+using std::pmr::polymorphic_allocator;
+
+void allocate_bytes_test() {
+    constexpr int N = 5;
+
+    polymorphic_allocator<> alloc{};
+
+    void* vp = alloc.allocate_bytes(sizeof(int) * N, alignof(int));
```
Comment (StephanTLavavej) [inline tests/std/tests/P0339R6_polymorphic_allocator/test.cpp:14]: This test is good; should there also be test coverage for the default function argument of `alignment = alignof(max_align_t)`?

File: tests/std/tests/P0339R6_polymorphic_allocator/test.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    alloc.deallocate_bytes(vp, N, alignof(int));
+}
+
+void allocate_object_test() {
+    constexpr int N = 10;
+
+    polymorphic_allocator<> alloc{};
+
+    int* arr = alloc.allocate_object<int>(N);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0339R6_polymorphic_allocator/test.cpp:38]: Similarly, should there be additional test coverage for the default function argument of `n = 1`?

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]
Adding 2 comments that somehow got lost from the initial review.

File: stl/inc/xpolymorphic_allocator.h [hunk 6]
```diff
@@ -13,6 +13,10 @@
 #include <utility>
 #include <xmemory>
 
+#if _HAS_CXX20
+#include <memory>
+#endif // _HAS_CXX20
```
[the remark this answers is not in the document]
Comment (onihusube) [reply stl/inc/xpolymorphic_allocator.h:18]: Only `std::destroy_at()`.
I think that would be preferable.
But I'm not sure how to move it (mainly, where?)

File: stl/inc/xpolymorphic_allocator.h [hunk 7]
```diff
@@ -186,7 +190,11 @@ namespace pmr {
     }
 
     // CLASS TEMPLATE polymorphic_allocator
+#if _HAS_CXX20
+    template <class _Ty = byte>
```
[the remark this answers is not in the document]
Comment (onihusube) [reply stl/inc/xpolymorphic_allocator.h:194]: I prefer Option B, too.

File: stl/inc/xpolymorphic_allocator.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/xpolymorphic_allocator.h:18]: That's a good question - we generally try to follow the Standard's order of declarations when there isn't a reason otherwise - however that doesn't provide much of a guide here. While trying to look for related machinery, I realized that #1313 (created today) is changing `std::destroy_at()`'s definition - so to avoid merge conflicts, let's leave this code as-is for now. We can refactor to improve throughput in the future.

[the remark this answers is not in the document]
Comment (onihusube) [reply stl/inc/xpolymorphic_allocator.h:18]: I see. Yes, sir!

[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/xpolymorphic_allocator.h:18]: I think the official response is one of "Meow", "Hiss" and "Purr"

**Review by StephanTLavavej — APPROVED** [review]
Looks great, thanks! I pushed a couple of small changes to the test:

* Use `std::uintptr_t` for modulo. (Unsigned modulo is easier to reason about; signed modulo was notoriously tricky although [[expr.mul]](https://eel.is/c++draft/expr.mul) was eventually simplified in C++17 if I recall correctly.)
* Update include directives:
  + Need `<cstdint>` for `std::uintptr_t`.
  + Need `<limits>`, not `<numeric>`, for `std::numeric_limits`.
  + Need `<new>` for `std::bad_alloc` and `std::bad_array_new_length`.

I'll move your PR forward to Final Review - then we'll just need a second STL maintainer to sign off (they tend to be active during Redmond's daytime, unlike me :joy_cat:), and then your feature will be ready to merge! :tada:

File: stl/inc/xpolymorphic_allocator.h [hunk 8]
```diff
@@ -229,6 +237,49 @@ namespace pmr {
             _Uses_allocator_construct(_Ptr, _Al, *this, _STD forward<_Types>(_Args)...);
         }
 
+#if _HAS_CXX20
+
```
Comment (CaseyCarter) [inline stl/inc/xpolymorphic_allocator.h:241]:
```suggestion
```

File: stl/inc/xpolymorphic_allocator.h [hunk 9]
```diff
@@ -229,6 +237,49 @@ namespace pmr {
...
+            _CATCH_END
+            return _Ptr;
+        }
+
+        template <class _Uty>
+        void delete_object(_Uty* const _Ptr) noexcept /* strengthened */ {
+            _STD destroy_at(_Ptr);
+            this->deallocate_object(_Ptr);
+        }
+
```
Comment (CaseyCarter) [inline stl/inc/xpolymorphic_allocator.h:280]:
```suggestion
```

File: stl/inc/xpolymorphic_allocator.h [hunk 10]
```diff
@@ -229,6 +237,49 @@ namespace pmr {
             _Uses_allocator_construct(_Ptr, _Al, *this, _STD forward<_Types>(_Args)...);
         }
 
+#if _HAS_CXX20
```
Comment (CaseyCarter) [inline stl/inc/xpolymorphic_allocator.h:240]: `construct` appears after the declarations of these new functions in [mem.poly.allocator.class.general]/2. Could we please move it after the `#endif` so everything is in synopsis order?

File: stl/inc/xpolymorphic_allocator.h [hunk 11]
```diff
@@ -229,6 +237,49 @@ namespace pmr {
...
+        }
+
+        void deallocate_bytes(void* const _Ptr, const size_t _Bytes,
+            const size_t _Align = alignof(max_align_t)) noexcept /* strengthened */ {
+            _Resource->deallocate(_Ptr, _Bytes, _Align);
+        }
+
+        template <class _Uty>
+        _NODISCARD __declspec(allocator) _Uty* allocate_object(_CRT_GUARDOVERFLOW const size_t _Count = 1) {
+            void* const _Vp = this->allocate_bytes(_Get_size_of_n<sizeof(_Uty)>(_Count), alignof(_Uty));
```
Comment (CaseyCarter) [inline stl/inc/xpolymorphic_allocator.h:254]: We don't conventionally use extraneous `this->` before member names, reserving that syntax for when we need to make lookups into dependent base classes. (Also on 260, 265, 267, 269, and 278.)

File: stl/inc/xpolymorphic_allocator.h [hunk 12]
```diff
@@ -229,6 +237,49 @@ namespace pmr {
...
+        }
+
+        template <class _Uty>
+        void deallocate_object(_Uty* const _Ptr, const size_t _Count = 1) noexcept /* strengthened */ {
+            this->deallocate_bytes(_Ptr, _Count * sizeof(_Uty), alignof(_Uty));
+        }
+
+        template <class _Uty, class... _Types>
+        _NODISCARD __declspec(allocator) _Uty* new_object(_Types&&... _Args) {
+            _Uty* _Ptr = this->allocate_object<_Uty>();
```
Comment (CaseyCarter) [inline stl/inc/xpolymorphic_allocator.h:265]: `_Ptr` could be `const` here.

File: stl/inc/xpolymorphic_allocator.h [hunk 13]
```diff
@@ -229,6 +237,49 @@ namespace pmr {
...
+            _CATCH_ALL
+            this->deallocate_object(_Ptr);
+            _RERAISE;
+            _CATCH_END
+            return _Ptr;
+        }
+
+        template <class _Uty>
+        void delete_object(_Uty* const _Ptr) noexcept /* strengthened */ {
+            _STD destroy_at(_Ptr);
```
Comment (CaseyCarter) [inline stl/inc/xpolymorphic_allocator.h:277]: We could equivalently use `_Destroy_in_place(*_Ptr)` here, and remove the `<memory>` include.

File: tests/std/tests/P0339R6_polymorphic_allocator/test.cpp [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        alloc.construct(arr + i, i);
+    }
+
+    for (int i = 0; i < N; ++i) {
+        assert(arr[i] == i);
+    }
+
+    for (int i = 0; i < N; ++i) {
+        std::destroy_at(arr + i);
+    }
```
Comment (CaseyCarter) [inline tests/std/tests/P0339R6_polymorphic_allocator/test.cpp:33]: Why not `std::destroy(arr, arr + N);`? (And on 57.)

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

**Comment by SuperWig** [comment]
Shouldn't that use the macro?

**Comment by onihusube** [comment]
`_Destroy_in_place`?  
Isn't it necessary because it is a reserved word?

**Comment by SuperWig** [comment]
No, you've used `std::destroy` which should be `_STD destroy`. The macro expands to `::std::` and is needed to prevent hijacking via ADL.

So the only real difference is `::` at the beginning but the macro should be used for consistency and being immensely shorter.

**Comment by AdamBucior** [comment]
> No, you've used `std::destroy` which should be `_STD destroy`. The macro expands to `::std::` and is needed to prevent hijacking via ADL.
> 
> So the only real difference is `::` at the beginning but the macro should be used for consistency and being immensely shorter.

But he used it in test.

**Comment by SuperWig** [comment]
... ignore me.

**Review by StephanTLavavej — APPROVED** [review]
Looks good; I am not sure why the tests timed out so I'll rerun them.

**Comment by StephanTLavavej** [comment]
Status update: I'm planning to merge this tomorrow, or Friday at the latest - we've accumulated a few other PRs that are ready to merge, and we like to batch them up (as they have to be simultaneously merged here and our MS-internal repo, and that involves a bit of manual work).

**Comment by StephanTLavavej** [comment]
Thanks again for implementing this feature, and congratulations on your first microsoft/STL commit! :tada: :smiley_cat: :rocket:
